### ISSUE TITLE
[pitchshifter] It seems that the "true" low frequencies will be lost if a target bin receives contributions from multiple source bins

### ISSUE THREAD

**Issue by stephanemagnenat** [nwoeanhinnogaehr/pvoc-plugins#1]
In [line 42](https://github.com/nwoeanhinnogaehr/pvoc-plugins/blob/dad17ef2450f8d5e6dfbbc4814e312f475149d2f/src/plugins/pitchshifter.rs#L42) the frequency is overwritten, while in [line 43](https://github.com/nwoeanhinnogaehr/pvoc-plugins/blob/dad17ef2450f8d5e6dfbbc4814e312f475149d2f/src/plugins/pitchshifter.rs#L43) the amplitude is added. If I understand this code correctly, this means that if multiple source bins are written to the same target bin, due to the pitch being lowered, then only the highest "true" frequency is written.

I am aware that the original source code in the [article](http://blogs.zynaptiq.com/bernsee/pitch-shifting-using-the-ft/) does this as well, but this feels wrong to me. I believe it is probably better to do some form of averaging, for example a weighted average of the incoming frequencies using their corresponding amplitudes. This will consume more processing power, but it should be possible to do something not too heavy, for example using an arena allocator with a capacity equal to the number of processed bins.

**Comment by nwoeanhinnogaehr** [comment]
Good point! I agree that a weighted average of frequencies would probably do a better job. Feel free to send a pull request if you want. Otherwise, I will get around to it eventually.

**Comment by nwoeanhinnogaehr** [comment]
So I just tried implementing this, and unexpectedly it actually sounds kind of hilariously bad. Like there is some kind of drunk vibrato applied independently to all of the frequency bins.

I think maybe using the frequency from the incoming bin with the largest amplitude may be better. Or maybe the median frequency or unweighted average. I'll try out a few things. In any case these improvements will probably not get backported here but instead go to my new plugin that I'm working on which obsoletes these.
